beginner_source/profiler_tutorial.py

```diff
@@ -0,0 +1,203 @@
+# Performance debugging using Profiler
+# ~~~~~~~~~~~~~~~~~
+#
+# Profiler can be useful to identify performance bottlenecks in your
+# models. In this example, we build a custom module that performs two
+# sub-tasks:
+#
+# - a linear transformation on the input, and
+# - use the transformation result to get indices on a mask tensor.
+#
+# We wrap the code for each sub-task in separate labelled context managers using
+# ``profiler.record_function("label")``. In the profiler output, the
+# aggregate performance metrics of all operations in the sub-task will
+# show up under its corresponding label.
+#
+#
+# Note that using Profiler incurs some overhead, and is best used only for investigating
+# code. Remember to remove it if you are benchmarking runtimes.
+#
+
+class MyModule(nn.Module):
+    def __init__(self, in_features: int, out_features: int, bias: bool = True):
+        super(MyModule, self).__init__()
+        self.linear = nn.Linear(in_features, out_features, bias)
+
+    def forward(self, input, mask):
+        with profiler.record_function("LINEAR PASS"):
+            out = self.linear(input)
+
+        with profiler.record_function("MASK INDICES"):
+            threshold = out.sum(axis=1).mean().item()
+            hi_idx = np.argwhere(mask.cpu().numpy() > threshold)
+
+        return out, hi_idx
+
+
+######################################################################
+# Profile the forward pass
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# We initialize random input and mask tensors, and the ``CustomLinear`` module.
+#
+# Before we run the profiler, we warm-up CUDA to ensure accurate
+# performance benchmarking. We wrap the forward pass of our module in the
+# ``profiler.profile`` context manager. The ``with_stack=True`` parameter appends the
+# file and line number of the operation in the trace.
+#
+
+model = MyModule(500, 10).cuda()
+input = torch.rand(128, 500).cuda()
+mask = torch.rand((500, 500, 500), dtype=torch.double).cuda()
+
+# warm-up
+model(input, mask)
+
+with profiler.profile(with_stack=True, profile_memory=True) as prof:
+    out, idx = model(input, mask)
+
+
+######################################################################
+# Pretty print profiler results
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# Finally, we print the profiler results. ``profiler.key_averages``
+# aggregates the results by operator name, and optionally by input
+# shapes and/or stack trace events.
+# Grouping by input shapes is useful to identify which tensor shapes
+# are utilized by the model.
+#
+# Here, we use ``group_by_stack_n=5`` which aggregates runtimes by the
+# operation and its traceback (truncated to the most recent 5 events), and
+# display the events in the order they are registered. The table can also
+# be sorted by passing a ``sort_by`` argument (refer to the
+# `docs <https://pytorch.org/docs/stable/autograd.html#profiler>`__ for
+# valid sorting keys).
+#
+
+print(prof.key_averages(group_by_stack_n=5).table())
```

Comment:
when outputting stats here and in other places - let's use `table(sort_by=<metric>, row_limit=X)`

where X - is up to you, just to comfortably see the interesting parts, -1 means 'all rows'
and metric - for memory it is "cpu_memory_usage" and "cuda_memory_usage"